python/tk_multi_breakdown2/file_item_model.py

```diff
@@ -1270,48 +1270,41 @@ def _get_published_files_for_items(self, file_items, data_retriever=None):
 
     def _get_published_files_mapping(self, published_file_data):
         """
-        Return a mapping of published files by their entity, name, task and type.
+        Return a mapping of Published Files by their fields specified in app settings.
 
-        The published file data passed in is a list of published file data (dictionaries). A
-        mapping is created where the published files are indexed by their entity, name, task,
-        and published file type.
+        For example, if publish_history_group_by_fields is set to ["project", "entity", "task", "name"],
+        then the returned dictionary will be a nested dictionary mapping published files by their
+        project, entity, task, and name.
+
+        For each level, if the value is a dictionary, then the dictionary will be mapped by the
+        dictionary's "type" and "id" values, else the value will be used as the key.
 
         :param published_file_data: The list of published file data to map.
         :type published_file_data: List[dict]
 
         :return: The dictionary mapping for published files.
         :rtype: dict
         """
-
+        group_by_fields = self._app.get_setting("publish_history_group_by_fields")
         published_files_mapping = {}
 
         for pf_data in published_file_data:
-            name = pf_data.get("name")
-
-            if not pf_data.get("entity"):
-                entity_id = None
-                entity_type = None
-            else:
-                entity_id = pf_data["entity"]["id"]
-                entity_type = pf_data["entity"]["type"]
-
-            if not pf_data.get("task"):
-                task_id = None
-            else:
-                task_id = pf_data["task"]["id"]
-
-            if not pf_data.get("published_file_type"):
-                pf_type_id = None
-            else:
-                pf_type_id = pf_data["published_file_type"]["id"]
-
-            published_files_mapping.setdefault(entity_type, {}).setdefault(
-                entity_id, {}
-            ).setdefault(task_id, {}).setdefault(pf_type_id, {}).setdefault(
-                name, []
-            ).append(
-                pf_data
-            )
+            current_level = published_files_mapping
+
+            # Iterate over fields and navigate or create nested dictionary structure.
+            for index, field in enumerate(group_by_fields):
+                field_value = pf_data.get(field)
+                if isinstance(field_value, dict):
+                    field_value = (field_value.get("type"), field_value.get("id"))
+
+                # If it's last field in the loop, ensure the current_level for
+                # that key is a list and append to it.
+                if index == len(group_by_fields) - 1:
```

Comment:
We can't use a defaultdict because of that, right? May be add a comment mentioning that?